Assets/Mirror/Runtime/NetworkAuthenticator.cs

```diff
@@ -0,0 +1,71 @@
+﻿using System;
+using UnityEngine;
+using UnityEngine.Events;
+
+namespace Mirror
+{
+    /// <summary>
+    /// Unity Event for the NetworkConnection
+    /// </summary>
+    [Serializable] public class UnityEventConnection : UnityEvent<NetworkConnection> { }
+
+    /// <summary>
+    /// Base class for implementing component-based authentication during the Connect phase
+    /// </summary>
+    [AddComponentMenu("Network/NetworkAuthenticator")]
+    [HelpURL("https://mirror-networking.com/xmldocs/articles/Concepts/Authentication.html")]
+    public class NetworkAuthenticator : MonoBehaviour
+    {
+        [Header("Event Listeners (optional)")]
+
+        /// <summary>
+        /// Notify subscribers on the server when a client is authenticated
+        /// </summary>
+        [Tooltip("Mirror has an internal subscriber to this event. You can add your own here.")]
+        public UnityEventConnection OnServerAuthenticated = new UnityEventConnection();
```

Comment:
what do we need those events for?

Reply:
this is where the magic happens.

The authenticator invokes this event to let the NetworkManager know that the connection has been successfully authenticated.

The nework manager subscribes to this and waits until the authenticator approves the connection before doing anything else

Reply:
note that the OnValidate below will only work if on networkmanager gameobject